publicsuffix/publicsuffix.go

```diff
@@ -195,43 +195,23 @@ Scanning:
 
 // Find and returns the most appropriate rule for the domain name.
 func (l *List) Find(name string, options *FindOptions) *Rule {
-	var bestRule *Rule
-
 	if options == nil {
 		options = DefaultFindOptions
 	}
 
-	for _, r := range l.selectRules(name, options) {
-		if r.Type == ExceptionType {
-			return &r
-		}
-		if bestRule == nil || bestRule.Length < r.Length {
-			bestRule = &r
-		}
-	}
-
-	if bestRule != nil {
-		return bestRule
-	}
-
-	return options.DefaultRule
-}
-
-func (l *List) selectRules(name string, options *FindOptions) []Rule {
-	var found []Rule
-
-	// In this phase the search is a simple sequential scan
-	for _, rule := range l.rules {
-		if !rule.Match(name) {
-			continue
+	for {
+		rule, ok := l.rules[name]
```

Comment:
There is a subtle but important implication in this change, and I want to add a note here mostly as a reminder.
This change will assume that `*.foo` and `foo` won't exist in the same definition.

Please see the main ticket for further information.